arch: arm: cortex_m: scb: Consolidate SCB APIs and add backup/restore functions (2nd attempt)

mathieuchopstm opened the pull request:
CMSIS 5 had a different name for the SHP/SHPR registers depending on which Cortex-M was used. This inconsistency is fixed in CMSIS 6, which is the new default for Cortex-M in Zephyr.

Remove the conditional and always use SHPR in the SCB backup/restore code.

Should fix `example-application` breakage as seen here https://github.com/zephyrproject-rtos/example-application/actions/runs/16510871831

msmttchr committed:
arch: arm: cortex_m: Add API for SCB save and restore

Add new API to save and restore SCB context. This is typically useful when
entering and exiting suspend-to-RAM low-power modes.

The scb_context_t and the backup/restore functions are designed to only
handle SCB registers that are:
- Mutable: Their values can be changed by software.
- Configurable: They control system behavior or features.
- Stateful: Their values represent a specific configuration that an
            application might want to preserve and restore.

Registers excluded from backup/restore are:
1. CPU/feature identification registers
	Motivation: These registers are fixed in hardware and read-only.
2. ICSR (Interrupt Control and State Register)
	Motivation: Most bits of ICSR bits are read-only or write-only
	and represent volatile system state. STTNS is the only read-write
	field and could be considered part of the system state, but it is
	only present on certain ARMv8-M CPUs, and Zephyr does not use it.
3. CFSR (Configurable Fault Status Register)
   HFSR (HardFault Status Register)
   DFSR (Debug Fault Status Register)
   AFSR (Auxiliary Fault Status Register)
   MMFAR (MemManage Fault Address Register)
   BFAR (BusFault Address Register)
	Motivation: These registers are read/write-one-to-clear and
	contain only fault-related information (which is volatile).

Co-authored-by: Mathieu Choplain <mathieu.choplain@st.com>

Signed-off-by: Michele Sardo <msmttchr@gmail.com>
Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm commented:
For some reason, CI adds *both* CMSIS 5 and CMSIS 6 to include path in this order. Because CMSIS 5 comes first, its outdated definitions are taken and build is broken:
<img width="1603" height="284" alt="image" src="https://github.com/user-attachments/assets/256c09c1-3f8b-4c95-81f5-86beaca0838f" />

On the other hand, [`example-application` CI picks up only CMSIS 6 as expected](https://github.com/zephyrproject-rtos/example-application/actions/runs/16510871831):
<img width="1607" height="261" alt="image" src="https://github.com/user-attachments/assets/ebb1e5a5-1197-430a-8ab2-fcb4555a5745" />

mathieuchopstm commented:
Keeping this open for re-introduction of https://github.com/zephyrproject-rtos/zephyr/pull/93161 when CI issues are fixed.

fabiobaltieri commented:
sure just trying to figure why both cmsis path are included, doubt it's intentional, the example application does not checkout cmsis, it only has cmsis_6 in manifest so it works there

fabiobaltieri commented:
would be also nice to understand why CI did not catch it, has to be an interaction with something else that got merged in the meantime

mathieuchopstm commented:
> would be also nice to understand why CI did not catch it, has to be an interaction with something else that got merged in the meantime --- oh I see now, it only works with both cmsis packages

More specifically, CI has both `modules/hal/cmsis` and `modules/hal/cmsis_6` in include path and - more critically - `cmsis` (CMSIS 5) comes first in include order.

@msmttchr's PR was written with CMSIS 5 in mind so CI was all but happy :slightly_smiling_face:

mathieuchopstm commented:
Pushed cleaned-up code from #93161 that should be OK for integration once CI uses CMSIS 6.

mathieuchopstm committed:
doc: release: release notes-4.3: document new ARM C-M save/restore APIs

Add node about new architecture APIs to save and restore ARM Cortex-M
system state.

Signed-off-by: Mathieu CHOPLAIN <mathieu.choplain@st.com>

mathieuchopstm commented:
Will rebase + add release notes in a moment...

mathieuchopstm commented:
Waiting on #94162 now
